plan/new_physical_plan_builder.go

```diff
@@ -101,172 +101,185 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
 	return matchOffsets
 }
 
+func (p *LogicalJoin) constructIndexJoin(innerJoinKeys, outerJoinKeys []*expression.Column, outerIdx int) *PhysicalIndexJoin {
+	var rightConds, leftConds expression.CNFExprs
+	if outerIdx == 0 {
+		rightConds = p.RightConditions.Clone()
+		leftConds = p.LeftConditions.Clone()
+	} else {
+		rightConds = p.LeftConditions.Clone()
+		leftConds = p.RightConditions.Clone()
+	}
+	join := PhysicalIndexJoin{
+		outerIndex:      outerIdx,
+		LeftConditions:  leftConds,
+		RightConditions: rightConds,
+		OtherConditions: p.OtherConditions,
+		Outer:           p.JoinType != InnerJoin,
+		OuterJoinKeys:   outerJoinKeys,
+		InnerJoinKeys:   innerJoinKeys,
+		DefaultValues:   p.DefaultValues,
+	}.init(p.allocator, p.ctx, p.children[outerIdx], p.children[1-outerIdx])
+	return join
+}
+
 // convertToIndexJoin will generate index join by required properties and outerIndex. OuterIdx points out the outer child,
 // because we will swap the children of join when the right child is outer child.
 // First of all, we will extract the join keys for p's equal conditions. If the join keys can match some of the indices or pk
 // column of inner child, we can apply the index join. Then we convert the inner child to table scan or index scan explicitly.
-func (p *LogicalJoin) convertToIndexJoin(prop *requiredProp, outerIdx int) (task, error) {
-	outerChild := p.children[outerIdx].(LogicalPlan)
+func (p *LogicalJoin) getIndexJoinByOuterIdx(outerIdx int) PhysicalPlan {
 	innerChild := p.children[1-outerIdx].(LogicalPlan)
 	var (
-		useTableScan  bool
 		usedIndexInfo *model.IndexInfo
-		rightConds    expression.CNFExprs
-		leftConds     expression.CNFExprs
-		innerTask     task
 		innerJoinKeys []*expression.Column
 		outerJoinKeys []*expression.Column
 	)
-	outerTask, err := outerChild.convert2NewPhysicalPlan(&requiredProp{taskTp: rootTaskType})
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
 	if outerIdx == 0 {
-		rightConds = p.RightConditions.Clone()
-		leftConds = p.LeftConditions.Clone()
 		outerJoinKeys = p.LeftJoinKeys
 		innerJoinKeys = p.RightJoinKeys
 	} else {
-		rightConds = p.LeftConditions.Clone()
-		leftConds = p.RightConditions.Clone()
 		innerJoinKeys = p.LeftJoinKeys
 		outerJoinKeys = p.RightJoinKeys
 	}
-	for {
-		switch x := innerChild.(type) {
-		case *DataSource:
-			indices, includeTableScan := availableIndices(x.indexHints, x.tableInfo)
-			if includeTableScan && len(innerJoinKeys) == 1 {
-				pkCol := x.getPKIsHandleCol()
-				if pkCol != nil && innerJoinKeys[0].Equal(pkCol, nil) {
-					useTableScan = true
-				}
+	if x, ok := innerChild.(*DataSource); ok {
```

Comment:
Why not consider Selection? If we only consider DataSource there, it is better to write code like this:
x, ok := innerChild.(*DataSource)
if !ok {
return nil
}